### PR TITLE
[frieren] DrivAerML Cross-Attention Bridge Between Surface and Volume Tokens

### PR DESCRIPTION
**Pull request by morganmcg1**
# [frieren] DrivAerML Cross-Attention Bridge Between Surface and Volume Tokens

## Hypothesis

The current Transolver uses a shared backbone that processes surface and volume tokens together. This means surface tokens must compete with volume tokens for attention budget. A cross-attention bridge allows explicit information exchange: surface tokens query volume tokens to understand the broader flow field context, and volume tokens query surface tokens to understand boundary conditions.

This mirrors how physical CFD works: boundary conditions (surface pressure/shear) are set by the no-slip and pressure conditions at the wall, and they directly drive the volume flow field. A cross-attention bridge models this causal structure explicitly.

**Architecture:** Keep the existing shared Transolver backbone but add a cross-attention layer between surface and volume feature streams at the final backbone layer before the output heads.

**Prediction:** Cross-attention will improve `volume_pressure_rel_l2_pct` significantly (volume predictions will better respect surface boundary conditions) and improve `wall_shear_*` (surface predictions will use global flow field context).

## Implementation Instructions

Add a cross-attention module to `model.py`:

```python
class CrossAttentionBridge(nn.Module):
    """
    Cross-attention between two feature streams.
    Query from one stream, Key/Value from the other.
    """
    def __init__(self, hidden_dim: int, n_heads: int = 4):
        super().__init__()
        self.n_heads = n_heads
        self.head_dim = hidden_dim // n_heads
        self.scale = self.head_dim ** -0.5
        
        # Surface queries volume
        self.surf_q = nn.Linear(hidden_dim, hidden_dim)
        self.vol_kv = nn.Linear(hidden_dim, hidden_dim * 2)
        self.surf_out = nn.Linear(hidden_dim, hidden_dim)
        
        # Volume queries surface
        self.vol_q = nn.Linear(hidden_dim, hidden_dim)
        self.surf_kv = nn.Linear(hidden_dim, hidden_dim * 2)
        self.vol_out = nn.Linear(hidden_dim, hidden_dim)
        
        self.surf_norm = nn.LayerNorm(hidden_dim)
        self.vol_norm = nn.LayerNorm(hidden_dim)

    def forward(self, surf_feat: torch.Tensor, vol_feat: torch.Tensor):
        """
        surf_feat: [B, N_surf, D]
        vol_feat: [B, N_vol, D]
        Returns: updated surf_feat, vol_feat
        """
        B, Ns, D = surf_feat.shape
        _, Nv, _ = vol_feat.shape
        H, Hd = self.n_heads, self.head_dim
        
        # Surface queries volume context
        q_s = self.surf_q(surf_feat).reshape(B, Ns, H, Hd).transpose(1, 2)
        kv_v = self.vol_kv(vol_feat).reshape(B, Nv, 2, H, Hd).permute(2, 0, 3, 1, 4)
        k_v, v_v = kv_v[0], kv_v[1]
        attn_sv = (q_s @ k_v.transpose(-2, -1)) * self.scale
        attn_sv = attn_sv.softmax(dim=-1)
        surf_cross = (attn_sv @ v_v).transpose(1, 2).reshape(B, Ns, D)
        surf_feat = self.surf_norm(surf_feat + self.surf_out(surf_cross))
        
        # Volume queries surface context
        q_v = self.vol_q(vol_feat).reshape(B, Nv, H, Hd).transpose(1, 2)
        kv_s = self.surf_kv(surf_feat).reshape(B, Ns, 2, H, Hd).permute(2, 0, 3, 1, 4)
        k_s, v_s = kv_s[0], kv_s[1]
        attn_vs = (q_v @ k_s.transpose(-2, -1)) * self.scale
        attn_vs = attn_vs.softmax(dim=-1)
        vol_cross = (attn_vs @ v_s).transpose(1, 2).reshape(B, Nv, D)
        vol_feat = self.vol_norm(vol_feat + self.vol_out(vol_cross))
        
        return surf_feat, vol_feat
```

**Integration in `SurfaceTransolver`:**
1. Add `self.cross_attn = CrossAttentionBridge(n_hidden, n_heads)` in `__init__`
2. After the final transformer layer, split the combined token sequence back into surface and volume tokens
3. Apply `surf_feat, vol_feat = self.cross_attn(surf_feat, vol_feat)`
4. Concatenate them back and pass to the output heads

**Memory note:** Cross-attention has O(N_surf * N_vol) memory cost. With 40k surface + 40k volume = 40k*40k attention matrix. Use chunked attention or reduce to cross-attention over pooled representations if OOM occurs.

**Memory-efficient alternative:** Pool volume tokens to 512 representative tokens (using slice pooling from Transolver), apply cross-attention over pooled representations, then broadcast back.

## Training Command

```bash
torchrun --nproc_per_node=4 train.py \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --no-use-ema --lr 3e-4 --weight-decay 1e-4 --batch-size 2 \
  --epochs 50 --lr-cosine-t-max 30 \
  --train-surface-points 40000 --train-volume-points 40000 \
  --kill-thresholds "500:train/loss<5,2000:val_primary/abupt_axis_mean_rel_l2_pct<30" \
  --wandb-group bengio-stream2-frieren --agent frieren
```

## Baseline Targets to Beat (AB-UPT Reference)

| Metric | AB-UPT Target |
|--------|--------------|
| `test_primary/surface_pressure_rel_l2_pct` | 3.82 |
| `test_primary/wall_shear_rel_l2_pct` | 7.29 |
| `test_primary/volume_pressure_rel_l2_pct` | 6.08 |
| `test_primary/wall_shear_x_rel_l2_pct` | 5.35 |
| `test_primary/wall_shear_y_rel_l2_pct` | 3.65 |
| `test_primary/wall_shear_z_rel_l2_pct` | 3.63 |
| `test_primary/abupt_axis_mean_rel_l2_pct` | ~4.51 |

**Primary checkpoint metric:** `val_primary/abupt_axis_mean_rel_l2_pct` (lower is better)

## Success Criteria

- Report which memory approach you used (full or pooled cross-attention)
- Report all `test_primary/*` metrics
- Compare `volume_pressure_rel_l2_pct` vs baseline (key hypothesis metric)
- Note any OOM issues and what batch size / point count you ended up using

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: **Important: Preemptive torch.compile warning** — Multiple students on this wave have reported a crash when `compile_model=True` (the default) hits validation for the first time. The crash is a known PyTorch 2.x Inductor bug in `tiling_utils.get_pw_red_splits` that fires when eval batches have different point-count shapes than training batches.

**If your experiment has not yet run its first validation pass:** Add `--no-compile-model` to your launch command immediately before you hit this bug. Throughput drops from ~20 it/s to ~10-11 it/s but training is stable.

**If your experiment is already past epoch 1 and has not crashed:** You may have gotten lucky with batch shape alignment. Monitor closely at the next validation pass, but you can continue.

No code changes needed — just add `--no-compile-model` to your train.py launch flags.

**Comment by morganmcg1** [comment]
STUDENT frieren:

## Status Update

Cleaned up a duplicate torchrun on the pod. Two concurrent torchrun jobs were running with the same cross-attention bridge configuration:
- **Active run `l23vz4md` (kept)**: started 11:11 UTC, currently at step ~20.5k, completed epoch 1 (val_primary/abupt_axis_mean_rel_l2_pct=18.58)
- **Duplicate `kdweou8c` (killed)**: started 11:43 UTC, was at step ~3.5k, no validations yet

The duplicate was sharing the same 4 GPUs, halving throughput on both. After killing the duplicate, GPU memory dropped from 23.5 GB → 11.8 GB per GPU.

## Active run summary

Run: `l23vz4md` (frieren/cross-attention-bridge, group: bengio-stream2-frieren)
- Implementation: pooled bidirectional cross-attention bridge with 256 soft-pool tokens per stream, applied after the final backbone block, before the output norm/heads. Uses tanh-gated zero-init residual so the bridge starts as identity.
- 4L / 256d / 4-head Transolver, no EMA, lr=3e-4, T_max=30, --no-compile-model
- Epoch 1 (step 17,816, took 31 min): val_primary/abupt_axis_mean_rel_l2_pct=18.58 (vs AB-UPT target ~4.51), surface_pressure=12.62, wall_shear=20.36, volume_pressure=12.14
- Train loss dropped 2.2 → 0.06 over the first epoch, gradient norms healthy (~1).

Will continue monitoring across remaining epochs and submit results when training completes.

**Comment by morganmcg1** [comment]
STUDENT frieren:

## Status: epoch 9 / 50, training healthy

Run `l23vz4md` (group: bengio-stream2-frieren) — pooled bidirectional cross-attention bridge with 256 soft-pool tokens per stream, tanh-gated zero-init residual, applied between final backbone block and output norm/heads.

Validation trajectory:
| Epoch | val_axis | surface_p | wall_shear | volume_p |
|------:|---------:|----------:|-----------:|---------:|
| 1 | 18.58 | 12.62 | 20.36 | 12.14 |
| 3 | 12.05 | 7.93  | 13.47 | 7.46  |
| 5 | 10.65 | 6.86  | 11.88 | 6.56  |
| 7 | 10.00 | 6.42  | 11.14 | 6.18  |
| 9 |  9.55 | 6.06  | 10.69 | **5.89** |
| AB-UPT target | 4.51 | 3.82 | 7.29 | 6.08 |

`volume_pressure_rel_l2_pct` (the headline hypothesis metric) is now **already below the AB-UPT target** at epoch 9 (5.89 vs 6.08). `surface_pressure` and `wall_shear` still ~1.5x off target — expecting further drop as the cosine schedule (T_max=30, eta_min=1e-6) anneals.

Each epoch ≈ 31 min on 4 GPUs (`--no-compile-model` to dodge the inductor bug); ~21h remaining for the full 50 epochs. Will report final test metrics when the run completes.

**Comment by morganmcg1** [comment]
ADVISOR: Excellent result — volume_pressure at 5.89% at epoch 9 is already below the AB-UPT target of 6.08%. This is the most promising result in the entire wave.

The convergence trajectory is very strong:
| Epoch | val_axis | surface_p | wall_shear | volume_p |
|------:|---------:|----------:|-----------:|---------:|
| 7 | 10.00 | 6.42 | 11.14 | 6.18 |
| 9 | 9.55 | 6.06 | 10.69 | **5.89** |
| AB-UPT target | 4.51 | 3.82 | 7.29 | 6.08 |

Surface pressure (6.06) is also within striking distance of the AB-UPT 6.08 volume target. With the cosine schedule still annealing (T_max=30, you're at epoch 9), there's strong potential for all metrics to continue falling.

Two things to focus on as you continue:
1. **Keep an eye on whether surface_pressure and wall_shear converge further at epochs 20-30** when the cosine schedule hits its steepest descent phase. That's when the big gains usually come.
2. **At the end of training, evaluate from the best validation checkpoint**, not the terminal epoch — that's our paper-facing number for comparison with AB-UPT.

Outstanding work. This cross-attention bridge between surface and volume tokens is clearly doing something useful for the volume representation.

**Comment by morganmcg1** [comment]
ADVISOR: Check-in at epoch 23 (escalation after 14 epochs of silence).

Pulled W&B run `l23vz4md` directly. State=running, step=412,097 (epoch 23/50). Per-epoch table:

| Epoch | abupt% | surf_p% | wall_sh% | vol_p% | wsy% | wsz% |
|---|---|---|---|---|---|---|
| 9 | (your last update) | | | 5.89 | | |
| 22 | **8.467** | 5.370 | 9.556 | 5.146 | ... | ... |
| 23 | 8.497 | 5.370 | 9.556 | 5.146 | ... | ... |

**Status:**
- vol_p=5.146% **still below the 6.08% target** — the cross-attention bridge is doing useful work for volume pressure.
- abupt=8.467% (best, epoch 22) is significantly worse than alphonse's `m9775k1v` at 7.324% (epoch 24).
- The early epoch-9 vol_p signal of 5.89% has held but did not improve much further.

This is a partial win at best — vol_p contribution is real, but the bridge isn't carrying the rest. **Please post the full per-epoch table and your interpretation NOW.** Do not wait until epoch 50. We need to decide whether to truncate early and harvest test metrics, or let it run.

Specifically: when did vol_p stabilize? Is the bridge cross-attention helping after the initial epochs, or saturating early? If you have logs of bridge attention weights or per-task loss decomposition, post them.

**Comment by morganmcg1** [comment]
ADVISOR: **ESCALATION #3 — frieren, cross-attention bridge.**

Your W&B run  shows: abupt=8.467%, vol_p=5.146% at epoch 23. Your run is the 4th-best result in the wave so far.

Estimate: by now (~01:38Z May 1, ~2h since last escalation) you should be around epoch 27/50. Expected completion ~09:00Z May 1.

This is your third escalation. If there is no student response by the time the run completes, I will evaluate and merge/close based on W&B data alone. If you have any results, questions, or modifications to report, please do so now.

**Comment by morganmcg1** [comment]
ADVISOR: **FORCED HARVEST — PR #85, frieren, cross-attention bridge.**

Three escalations have gone unanswered. I have pulled W&B run `l23vz4md` directly. Run state: `running`, current step ~656k (~ep35 of 50). The val trajectory is now in slow regression past ep30 and continued training is wasting GPU.

## Best-validation-epoch profile (from W&B history)

`l23vz4md` best at **step 552,326 (~ep30)**:

| Metric | Best (ep30) | Latest (~ep35) | AB-UPT target |
|---|---:|---:|---:|
| `val_primary/abupt_axis_mean_rel_l2_pct` | **8.1723** | 8.2276 | 4.51 |
| `val_primary/surface_pressure_rel_l2_pct` | 5.1988 | 5.2394 | 3.82 |
| `val_primary/volume_pressure_rel_l2_pct` | **4.9788** | 4.9833 | 6.08 ✓ beats |
| `val_primary/wall_shear_x_rel_l2_pct` | 7.8253 | 7.8587 | 5.35 |
| `val_primary/wall_shear_y_rel_l2_pct` | 10.9123 | 11.0212 | 3.65 |
| `val_primary/wall_shear_z_rel_l2_pct` | 11.9461 | 12.0353 | 3.63 |

vol_p (4.98%) clearly beats AB-UPT's target — the cross-attention bridge is doing useful work on the volume side. abupt is in the upper-mid of the wave. wall-shear axes are the bottleneck (consistent across the wave).

## Required actions now

1. **Stop the training run** (`l23vz4md`) — past best-val and slowly regressing.
2. **Reload the best-val checkpoint at step 552,326 / ep30**.
3. **Run test evaluation** and log all 6 `test_primary/*` metrics: `abupt_axis_mean_rel_l2_pct`, `surface_pressure_rel_l2_pct`, `volume_pressure_rel_l2_pct`, `wall_shear_x_rel_l2_pct`, `wall_shear_y_rel_l2_pct`, `wall_shear_z_rel_l2_pct`. These are MANDATORY per `target/program.md` — final metrics MUST come from `test_primary/*` after reloading the best-val checkpoint.
4. **Post the test_primary results in a comment on this PR.**
5. **Mark the PR `status:review`** when the test eval is logged.

If you do not respond within the next polling cycle I will mark this PR `status:review` myself based on the val-best W&B data. Please run the test eval — it gives a clean comparison against AB-UPT.

Also — there are four divergent v2 arm runs in the same project (`pdf0r65v`, `45dbkcb7`, `6yqbmce0`, `p4vr6v80`) showing abupt 172–185% at step ~33k. Were these arm experiments launched from this PR's branch? If yes, please clarify what configuration they tested. They appear catastrophically unstable and should be killed if still running.

**Comment by morganmcg1** [comment]
ADVISOR: **REVIEW — PR #85 frieren (cross-attention bridge between surface and volume)**

Val metrics (W&B run `l23vz4md`, best ~ep35):
- val_abupt = 8.1723% (baseline: 7.2091% — +0.96pp, does **not** beat baseline)
- vol_p = 4.979% ✓ (beats 6.08% target)
- surf_p = 5.199%

**Decision: Close.** Cross-attention between surface and volume token streams is an interesting architectural idea, but the 0.96pp deficit is significant. The cross-attention bridge adds considerable complexity and computational cost without the expected benefit.

**Lesson**: The current surface/volume encoding architecture already has sufficient implicit coupling through the shared loss. Explicit cross-attention may be causing over-coupling or optimization difficulties.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #85 because Closing: Cross-attention surface-volume bridge achieves val_abupt=8.17% — 0.96pp behind baseline. Adds architectural complexity without benefit. The current implicit coupling is sufficient..
